src/Common/ZooKeeper/CMakeLists.txt

```diff
@@ -2,9 +2,32 @@ include("${ClickHouse_SOURCE_DIR}/cmake/dbms_glob_sources.cmake")
 
 add_headers_and_sources(clickhouse_common_zookeeper .)
 
+# for clickhouse server
 add_library(clickhouse_common_zookeeper ${clickhouse_common_zookeeper_headers} ${clickhouse_common_zookeeper_sources})
+target_compile_definitions (clickhouse_common_zookeeper PRIVATE -DZOOKEEPER_LOG)
+target_link_libraries (clickhouse_common_zookeeper
+    PUBLIC
+        clickhouse_common_io
+        common
+    PRIVATE
+        string_utils
+)
+# To avoid circular dependency from interpreters.
+if (OS_DARWIN)
+    target_link_libraries (clickhouse_common_zookeeper PRIVATE -Wl,-undefined,dynamic_lookup)
+else()
+    target_link_libraries (clickhouse_common_zookeeper PRIVATE -Wl,--unresolved-symbols=ignore-all)
```

Comment:
This is questionable. If you forget to explicitly instantiate a template, for example, you will get obscure runtime error like this
```
clickhouse: symbol lookup error: clickhouse: undefined symbol: _ZN2DB9SystemLogINS_26TransactionsInfoLogElementEE3addERKS1_
```
and server will crash. It's really unexpected and misleading that linking error happens in runtime.

Reply:
And seems like it's not needed: #33844

Reply:
There is undefined reference to `DB::SystemLog<DB::ZooKeeperLogElement>::add`:

```
$ nm -D ./src/Common/ZooKeeper/libclickhouse_common_zookeeperd.so | fgrep SystemLog | c++filt                                                                       U DB::SystemLog<DB::ZooKeeperLogElement>::add(DB::ZooKeeperLogElement const&)
```

Your patch works, because default is not `-Wl,--unresolved-symbols=report-all`, so I guess shared build on OSX may fail now.

Let's try #33860 instead.